examples/src/main/kotlin/example/AxiosSearch.kt

```diff
@@ -35,6 +35,17 @@ external interface AxiosConfigSettings {
     var timeout: Number
 }
 
+external interface AxiosError {
+    var message: String
+}
+
+external interface AxiosResponse {
+    var data: dynamic
```

Comment:
Let's make `axios` return `Promise<AxiosResponse>` instead of `dynamic.

We can also make both `AxiosResponse` and `axios` a generic type so that `data` has the type of type parameter not `dynamic`:

```
external interface AxiosResponse<T> {
  // we can use `val` instead of `var` here because we never assign those values ourselves
  val data: T
  ...
}

@JsModule("axios")
external fun <T> axios(config: AxiosConfigSettings): Promise<AxiosResponse<T>>
``` 

Then you can add type for data:
```
external interface Data {
  val country: String
  val state: String,
  val city: String
}
```

and use `axios` like this:
```
// the type for `response` should be inferred authomatically
axios<Data>(config).then { response -> ... }.catch { error: AxiosError -> ... }
```

Reply:
I don't think it is a good idea to add more typing due to below reason:

1. The axios not only get response, but also have catch error branch.
So, external fun <T> axios(config: AxiosConfigSettings): Promise<AxiosResponse<T>> not correct.

2. When I try hide/ignore catch branch firstly.  The compile show below error:
How to resolve it?
Can you show example code? Thanks.

src\App\AxiosSearch\AxiosSearch.kt:92:29: error: no value passed for parameter 'onFulfilled'
        axios<Data>(config).then { response ->
                            ^
src\App\AxiosSearch\AxiosSearch.kt:94:48: error: unresolved reference: data
                zipResult = ZipResult(response.data.country, response.data.state, response.data.city)
                                               ^
src\App\AxiosSearch\AxiosSearch.kt:94:71: error: unresolved reference: data
                zipResult = ZipResult(response.data.country, response.data.state, response.data.city)
                                                                       ^
src\App\AxiosSearch\AxiosSearch.kt:94:92: error: unresolved reference: data
                zipResult = ZipResult(response.data.country, response.data.state, response.data.city)
                                                                                           ^
src\App\AxiosSearch\AxiosSearch.kt:96:34: error: unresolved reference: status
            console.log(response.status)
                                 ^
src\App\AxiosSearch\AxiosSearch.kt:97:34: error: unresolved reference: statusText
            console.log(response.statusText)

3. When I add back catch branch, it show below error:
src\App\AxiosSearch\AxiosSearch.kt:102:18: error: type inference failed: fun <S> catch(onRejected: (Throwable) -> S): Promise<S>
cannot be applied to
((AxiosError) -> Unit)

                .catch { error: AxiosError ->
                 ^
src\App\AxiosSearch\AxiosSearch.kt:102:24: error: type mismatch: inferred type is (AxiosError) -> Unit but (Throwable) -> Unit was expected
                .catch { error: AxiosError ->
                       ^
src\App\AxiosSearch\AxiosSearch.kt:102:26: error: expected parameter of type Throwable
                .catch { error: AxiosError ->

Reply:
First, you have to put the `{ response -> ... }` lambda in parens, because it's not actually the last argument of `then`:

```
axios<Data>(config).then({ response: AxiosResponse<Data> ->
  setState {
    zipResult = ZipResult(response.data.country, response.data.state, response.data.city)
    errorMessage = ""
  }
})
```

Second, the `Promise` builtin type in Kotlin expects the failure argument to be a [`Throwable`](http://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-throwable/index.html). For this example it's OK, because we only use the `message` field:
```
.catch { error ->
  setState {
    zipResult = ZipResult("", "", "")
    // the Elvis operator is needed because `Throwable::message` is nullable
    errorMessage = error.message ?: ""
  }
}
```